### PR TITLE
TDAP051-Add-Login-Functionality

### PR DESCRIPTION
**Pull request by amit-ib**
Added login functionality using API

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -24,22 +23,24 @@ function App() {
...
-  }, []);
-
-  const [userProfile, setuserProfile] = useState<UserModal | null>();
+    setIsLogedin(accessToken ? true : false);
+  }, [accessToken]);
 
+  const [userName, setuserName] = useState<string | null>(
+    window.localStorage.getItem("userName")
+      ? window.localStorage.getItem("userName")
+      : ""
```
Comment (akash-dhande-ib) [inline src/App.tsx:32]: Keep this logic separate for readability.

Comment (akash-dhande-ib) [reply src/App.tsx:32]: Also this should be `token`

File: src/App.tsx [hunk 2]
```diff
@@ -24,22 +23,24 @@ function App() {
   });
 
   useEffect(() => {
-    setIsLogedin(accessToken ? true : true);
-    console.log("isLogedin", isLogedin, accessToken);
-  }, []);
-
-  const [userProfile, setuserProfile] = useState<UserModal | null>();
+    setIsLogedin(accessToken ? true : false);
```
Comment (akash-dhande-ib) [inline src/App.tsx:26]: Token will be the one received from api.

File: src/assets/scss/components/_login.scss [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  text-align: center;
+  input {
+    margin-bottom: 10px;
+  }
+  button {
+    width: 100%;
+    &.primary {
+      margin-bottom: 10px;
+    }
+  }
```
Comment (akash-dhande-ib) [inline src/assets/scss/components/_login.scss:11]: Do we really need this, or can we make it generic.

File: src/assets/scss/ui/forms/_button.scss [hunk 4]
```diff
@@ -36,12 +36,24 @@
...
+    &.button-logout {
+      background: none !important;
+      border: 0px !important;
+      box-shadow: none !important;
+      color: $white-color !important;
+      font-size: 16px !important;
+      div {
+        display: none;
+      }
+    }
```
Comment (akash-dhande-ib) [inline src/assets/scss/ui/forms/_button.scss:57]: Do we really need this?

File: src/components/LoginForm.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    formState: { errors },
+  } = useForm<loginForm>({
+    defaultValues: {
+      email: "",
+      password: "",
+    },
+  });
+
+  const onSubmit: SubmitHandler<loginForm> = (data) => {
+    props.loginFunctionality(data);
```
Comment (akash-dhande-ib) [inline src/components/LoginForm.tsx:23]: ?? name

File: src/components/LoginForm.tsx [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        {...register("email", { required: true })}
+        placeholder="Please enter email"
+      />
+      {errors.email?.type === "required" && "Please enter email"}
+      <input
+        {...register("password", { required: true })}
+        type="string"
+        placeholder="Please enter password"
+      />
+      <Button type="submit" label="Login" className="button primary" />
```
Comment (akash-dhande-ib) [inline src/components/LoginForm.tsx:37]: Do we need to add the class `button`

File: src/components/shared/Login.jsx [hunk 7]
```diff
@@ -1,54 +1,93 @@
...
+      setShowLoginButton(false);
+      setShowLogoutButton(true);
+    } else {
+      setShowLoginButton(true);
+      setShowLogoutButton(false);
+    }
+  }, [props.isLogedin]);
+
+  const onLoginSuccess = (res) => {
+    console.log(res, "RES");
```
Comment (akash-dhande-ib) [inline src/components/shared/Login.jsx:23]: Debug

File: src/components/shared/Login.jsx [hunk 8]
```diff
@@ -1,54 +1,93 @@
 import { GoogleLogin, GoogleLogout } from "react-google-login";
 import { useState, useEffect } from "react";
+import axios from "axios";
+import LoginForm from "../LoginForm";
 
 const clientId = process.env.REACT_APP_CLIENTID;
 
 const Login = (props) => {
   const [showLoginButton, setShowLoginButton] = useState(true);
   const [showLogoutButton, setShowLogoutButton] = useState(false);
```
Comment (akash-dhande-ib) [inline src/components/shared/Login.jsx:10]: Do we need two different variable/state for Boolean

File: src/components/shared/Login.jsx [hunk 9]
```diff
@@ -1,54 +1,93 @@
...
+  const onLogoutSuccess = () => {
     props.setIsLogedin(false);
     window.localStorage.removeItem("accessToken");
-    setShowLoginButton(true);
-    setShowLogoutButton(false);
   };
 
+  const loginApiAuth = (data) => {
+    axios
+      .post("https://todo-node-api.vercel.app/api/login", {
```
Comment (akash-dhande-ib) [inline src/components/shared/Login.jsx:46]: Store the api path in environment file.

File: src/components/shared/Login.jsx [hunk 10]
```diff
@@ -1,54 +1,93 @@
...
-          buttonText="Logout"
-          onLogoutSuccess={onLogoutSuccess}
-        />
+        <span class="button-google">
+          <GoogleLogout
+            clientId={clientId}
+            buttonText="Logout"
+            onLogoutSuccess={onLogoutSuccess}
+            className="button-logout"
+          />
```
Comment (akash-dhande-ib) [inline src/components/shared/Login.jsx:68]: Do we still need this?

File: src/components/shared/Login.jsx [hunk 11]
```diff
@@ -1,54 +1,93 @@
...
+        </span>
       ) : null}
-    </div>
+
+      <div className="login-form">
+        {showLoginButton ? (
+          <div>
+            <LoginForm loginFunctionality={(data) => loginApiAuth(data)} />
```
Comment (akash-dhande-ib) [inline src/components/shared/Login.jsx:75]: We can keep the whole login thing in the same component.

File: src/App.tsx (hunk 2 above)
Comment (akash-dhande-ib) [inline src/App.tsx:26]: Keep the authentication information in store, so whole application can use it.

**Review by akash-dhande-ib — CHANGES_REQUESTED** [review]
Added several comments.

File: src/App.tsx (hunk 1 above)
Comment (amit-ib) [reply src/App.tsx]: Updated

File: src/App.tsx (hunk 2 above)
Comment (amit-ib) [reply src/App.tsx:22]: Updated

File: src/assets/scss/components/_login.scss (hunk 3 above)
Comment (amit-ib) [reply src/assets/scss/components/_login.scss:11]: Removed, this is not needed

File: src/assets/scss/ui/forms/_button.scss (hunk 4 above)
Comment (amit-ib) [reply src/assets/scss/ui/forms/_button.scss]: no, removed

File: src/components/LoginForm.tsx (hunk 5 above)
Comment (amit-ib) [reply src/components/LoginForm.tsx]: Updated

File: src/components/shared/Login.jsx (hunk 8 above)
Comment (amit-ib) [reply src/components/shared/Login.jsx]: Removed, as implemented in another way

File: src/components/shared/Login.jsx (hunk 9 above)
Comment (amit-ib) [reply src/components/shared/Login.jsx]: Updated

File: src/components/shared/Login.jsx (hunk 10 above)
Comment (amit-ib) [reply src/components/shared/Login.jsx]: No, removed

File: src/components/shared/Login.jsx (hunk 11 above)
Comment (amit-ib) [reply src/components/shared/Login.jsx]: Moved to same component

File: src/App.tsx (hunk 2 above)
Comment (amit-ib) [reply src/App.tsx:22]: Updated

File: src/App.tsx [hunk 12]
```diff
@@ -1,66 +1,59 @@
...
-        scope: "",
-      });
-    }
-    gapi.load("client:auth2", start);
-  });
+  const userName = window.localStorage.getItem("userName")
+    ? window.localStorage.getItem("userName")
+    : "";
+  const setIsLogedin = (bolval: boolean) => {
+    dispatch(loginTodoAction(bolval));
```
Comment (akash-dhande-ib) [inline src/App.tsx:18]: What is bolval?

File: src/App.tsx [hunk 13]
```diff
@@ -1,66 +1,59 @@
...
+            Welcome - {userName} |{" "}
+            <Button label="Logout" onClick={onLogoutSuccess} styles="link" />
           </span>
         )}
       </div>
       <div className="container">
         <div className="todo-container">
-          <h3 className="text-center">Todo List</h3>
+          {state.isLogedin ? (
+            <h3 className="text-center">Todo List</h3>
```
Comment (akash-dhande-ib) [inline src/App.tsx:43]: `<h3 className="text-center">` this code is redundent.

File: src/App.tsx (hunk 12 above)
Comment (akash-dhande-ib) [inline src/App.tsx:18]: We can manipulate the auth related state in login page itself. What do you think?

File: src/assets/scss/_header.scss [hunk 14]
```diff
@@ -2,22 +2,26 @@
...
-      div {
-        display: none;
-      }
-    }
+  .link {
+    color: $white-color;
+    font-size: 16px;
   }
+  // .button-login {
+  //   text-align: left;
```
Comment (akash-dhande-ib) [inline src/assets/scss/_header.scss:10]: Remove commented code

File: src/assets/scss/ui/forms/_button.scss [hunk 15]
```diff
@@ -36,12 +43,24 @@
...
-  button {
-    border-radius: 5px !important;
-    border: solid 1px #ccc !important;
-    padding-right: 10px !important;
-  }
-}
+// .button-google {
+//   text-align: center;
+//   padding-bottom: 10px;
+//   margin-top: 10px;
```
Comment (akash-dhande-ib) [inline src/assets/scss/ui/forms/_button.scss:49]: unused code

File: src/components/shared/Login.tsx [hunk 16]
```diff
@@ -0,0 +1,81 @@
+import React from "react";
```
Comment (akash-dhande-ib) [inline src/components/shared/Login.tsx:1]: You can move the Login component from shared to auth folder.

File: src/store/actionTypes.ts [hunk 17]
```diff
@@ -2,3 +2,4 @@ export const ADD_TODO = "ADD_TODO";
 export const DELETE_TODO = "DELETE_TODO";
 export const MARK_DONE_TODO = "MARK_DONE_TODO";
 export const EDIT_TODO = "EDIT_TODO";
+export const LOGIN_TODO = "LOGIN_TODO";
```
Comment (akash-dhande-ib) [inline src/store/actionTypes.ts:5]: LOGIN_TODO seems confusing.

File: src/store/todoReducer.tsx [hunk 18]
```diff
@@ -5,20 +5,29 @@ import {
   DELETE_TODO,
   MARK_DONE_TODO,
   EDIT_TODO,
+  LOGIN_TODO,
 } from "./actionTypes";
 //import { v4 as uuidv4 } from "uuid";
-export interface todoList {
+export interface reducerSate {
```
Comment (akash-dhande-ib) [inline src/store/todoReducer.tsx:11]: Naming

File: src/App.tsx (hunk 12 above)
Comment (amit-ib) [reply src/App.tsx:18]: Kindly provide more detail

File: src/App.tsx (hunk 13 above)
Comment (amit-ib) [reply src/App.tsx:43]: Updated

File: src/App.tsx (hunk 12 above)
Comment (amit-ib) [reply src/App.tsx]: Updated

File: src/assets/scss/_header.scss (hunk 14 above)
Comment (amit-ib) [reply src/assets/scss/_header.scss]: Removed

File: src/assets/scss/ui/forms/_button.scss (hunk 15 above)
Comment (amit-ib) [reply src/assets/scss/ui/forms/_button.scss]: Removed

File: src/components/shared/Login.tsx (hunk 16 above)
Comment (amit-ib) [reply src/components/shared/Login.tsx]: Done

File: src/store/actionTypes.ts (hunk 17 above)
Comment (amit-ib) [reply src/store/actionTypes.ts]: Updated

File: src/store/todoReducer.tsx (hunk 18 above)
Comment (amit-ib) [reply src/store/todoReducer.tsx]: Updated to statesModal

**Review by akash-dhande-ib — APPROVED** [review]

**Review by akash-dhande-ib — APPROVED** [review]

**Pull request merged**